riff-raff/app/deployment/actors.scala

```diff
@@ -115,209 +110,310 @@ class DeployMetricsActor extends Actor with Logging {
+    // otherwise let's see what children are valid to return
+    else {
+      // candidates are all successors not already executing or completing
+      val nextDeploymentCandidates = deploymentGraph.successorDeploymentNodes(deployment)
+      // now filter for only tasks whose predecessors are all completed
+      val nextDeployments = nextDeploymentCandidates.filter { deployment => (deploymentGraph.predecessors(deployment) -- completed).isEmpty }
+      if (nextDeployments.nonEmpty) {
+        Deployments(nextDeployments)
+      } else {
+        FinishPath()
+      }
+    }
+  }
+  protected[deployment] def markExecuting(deployment: DeploymentNode) = {
+    executing += deployment
+  }
+  protected[deployment] def markComplete(deployment: DeploymentNode) = {
+    executing -= deployment
+    completed += deployment
+  }
+  protected[deployment] def markFailed(deployment: DeploymentNode) = {
+    executing -= deployment
+    failed += deployment
+  }
+  def finishRootContext() = if (!rootContextClosed) {
+    rootContextClosed = true
+    DeployReporter.finishContext(rootReporter)
+  }
+  def failRootContext() = if (!rootContextClosed) {
+    rootContextClosed = true
+    DeployReporter.failContext(rootReporter)
+  }
+  def failRootContext(message: String, exception: Throwable) = if (!rootContextClosed) {
+    rootContextClosed = true
+    DeployReporter.failContext(rootReporter, message, exception)
+  }
+  private def cleanup = {
+    finishRootContext()
+    deployCoordinator ! DeployCoordinator.CleanupDeploy(record.uuid)
+    context.stop(self)
+  }
+
+  override def toString: String = {
+    s"""
+       |UUID: ${record.uuid.toString}
+       |#Tasks: ${allDeployments.size}
+       |#Executing: ${executing.mkString("; ")}
+       |#Completed: ${completed.size} Failed: ${failed.size}
+       |#Done: ${completed.size+failed.size}
+     """.stripMargin
+  }
+
+  def receive = {
+    case Start() =>
+      try {
+        self ! ContextCreated(createContext)
+        self ! StartDeployment()
+      } catch {
+        case NonFatal(t) =>
+          failRootContext("Preparing deploy failed", t)
+          cleanup
+      }
+
+    case ContextCreated(preparedContext) =>
+      deployContext = Some(preparedContext)
+
+    case StartDeployment() =>
+      runDeployments(firstDeployments)
+
+    case DeploymentCompleted(deployment) =>
+      log.debug("Deployment completed")
+      markComplete(deployment)
+      nextDeployments(deployment) match {
+        case Deployments(deployments) =>
+          runDeployments(deployments)
+        case FinishPath() =>
+        case FinishDeploy() =>
+          cleanup
+      }
+
+    case DeploymentFailed(deployment, exception) =>
+      log.debug("Deployment failed")
+      markFailed(deployment)
+      if (isExecuting) {
+        log.debug("Failed during deployment but others still running - deferring clean up")
+      } else {
+        cleanup
+      }
+
+    case Terminated(actor) =>
```

Comment:
I'm pretty sure you will receive this message in both the "child actor stopped cleanly by calling context.stop" and the "child actor exploded in a fireball" cases. So you might want to add a guard to only call `failRootContext` if `rootContextClosed == false`.